### PR TITLE
feat(memory): runtime rebuild operation with confirmed embedding change

### LINKED ISSUE #1314: feat(memory): runtime rebuild operation with confirmed embedding change
## Problem Statement

Changing `processing.embedding.base_url` or `processing.embedding.model` can leave Memory in a durable dead end. Avibe saves an `embedding_change_pending` marker, but reconcile only settles that marker when every vector-bearing surface is empty. With existing data the save fails, the UI rolls the candidate configuration back, and Restart engine remains fenced. The only documented escape today is Clear all, which deletes the user's memory.

The managed-child prerequisites are already merged into `dev`: artifact compatibility admission (#1307) and the role-aware `cascade rebuild` child capability (#1310). What is missing is the user-reachable recovery path: confirmed settings flow, runtime-owned rebuild orchestration over that child, an authenticated route, and the minimal Settings UI needed to confirm/retry.

## Solution

Ship a **narrow rebuild recovery path**, not a generic operations framework:

1. Confirmed embedding identity change on Memory settings PATCH.
2. Durable candidate + pending marker before any destructive work.
3. A **cross-process Memory config transaction** so UI save and Controller settlement cannot clobber each other with stale snapshots.
4. `MemoryRuntime.rebuild()` as one retained/shielded task (same shape as existing restart): the HTTP request **waits for the final result**.
5. Minimal Settings UI for confirm / retry / `rebuild_required` warning.

Markdown is preserved. Old and new vector spaces are never mixed. Failure leaves a truthful Retry path and never silently escalates to factory reset.

## User Stories

1. As an Avibe user, I want to change embedding provider/model without losing markdown memory content.
2. As an Avibe user, I want an unconfirmed identity-changing save rejected with a clear rebuild-required error.
3. As an Avibe user, I want the confirmed candidate saved durably before destructive work so a crash retries with the same config.
4. As an Avibe user, I want a failed rebuild to keep the confirmed configuration (no post-confirm rollback).
5. As an Avibe user, I want an empty root to settle without launching a child.
6. As an Avibe user, I want an API-key correction under a pending marker to feed the next retry.
7. As an Avibe user, I want API-key-only changes with no pending marker to keep ordinary save behavior.
8. As an Avibe user, I want concurrent Memory mutations refused with a distinct in-progress error while rebuild runs.
9. As an Avibe user, I want a duplicate rebuild request to join the running task rather than fork.
10. As an Avibe user, I want boot with a pending marker to keep claims fenced and the sidecar down, with Retry exposed and no automatic destructive work.
11. As an Avibe user, I want root-busy reported via a stable public error code and never steered toward factory reset.
12. As an operator, I want status to expose only a minimal rebuild failure/pending projection — no secrets, job IDs, history, or progress endpoints.

### Confirmed settings write

- `PATCH /api/memory/settings` remains the only candidate-configuration write surface.
- Optional exact boolean `confirm_rebuild`.
- Unconfirmed identity change → `409 memory_embedding_rebuild_required`; **nothing saved** (including apparently empty roots, to avoid check-then-save races).
- Confirmed save persists candidate + pending marker **before** scheduling rebuild. No rollback after confirmation.
- Settings GET exposes read-only `rebuild_required` while the marker is pending.
- API-key-only, no marker: ordinary save/reconcile/rollback.
- API-key-only under marker: update candidate, return saved-but-`rebuild_required`, do not reconcile or clear the fence.

### Cross-process Memory config transaction (required)

UI and Controller run in different processes. Process-local UI locks and `CONFIG_LOCK` are **not** a cross-process transaction. Without an explicit narrow protocol, a concurrent Controller write-back can overwrite a freshly saved candidate or clear the marker.

Define a **narrow Memory config transaction**, not a general multi-writer config system:

- Candidate + pending marker are written as one durable unit visible to both processes.
- Controller settlement (clear marker / activate candidate) and UI save of a new candidate must not lose the last confirmed write.
- Stale Controller snapshots must not clobber a newer confirmed candidate/marker.
- After a successful rebuild settles the marker and **before** sidecar activation, update the Controller/restart candidate snapshot so activation uses the settled config.
- Cover with an explicit **stale save / rollback interleaving** test at the routes + runtime boundary (hermetic; no live service).

Do **not** invent a generic distributed lock service or operation journal.

### Rebuild orchestration

- Deep module is `MemoryRuntime`. Reuse the existing retained/shielded task + shutdown-join pattern from `restart()` (`_restart_task`).
- Add `MemoryRuntime.rebuild()` and `_rebuild_task`. Request waits for the final result (200 with closed outcome, or structured error). **No 202, no polling, no job IDs, no generic Controller operation projection, no reserved `factory_reset` kind.**
- At most one rebuild task; join-on-duplicate; other Memory mutations see `memory_operation_in_progress` (or the existing clear/maintenance conflict code if that is already the chokepoint — prefer one stable public code, document it in this issue’s acceptance).
- Before any destructive stop: validate pinned artifact usability and that the candidate embedding endpoint is complete enough to rebuild a non-empty root. Fail closed with no destructive work when validation fails.
- Order under existing reconcile/destructive lifecycle:
  1. claim fence + definitive quiesce (no force bypass)
  2. stop sidecar with proven process-group death
  3. strict-empty fast path → synthetic `completed_empty` (no child)
  4. otherwise launch pinned-artifact `cascade rebuild --yes` via the merged role-aware child with bounded timeout
  5. map closed child results; settle marker only on `completed` / `completed_empty`
  6. clear marker durably, refresh Controller/restart candidate snapshot, then destructive-cutover start (skip ordinary healthy-replacement endpoint preflight)
- Closed result set: `completed | completed_empty | root_busy | interrupted | timed_out | failed`.
- `root_busy` (EverOS exit 3) maps to a **stable public error code** (e.g. `memory_rebuild_root_busy`) and must never mention factory reset.
- Indeterminate data state fails closed.
- Boot with pending marker: reconcile ownership first, keep claims fenced, no sidecar, expose retry projection; never auto-start rebuild.
- Health after enabled startup: existing health projection only — operational warning vs failed-row data-quality warning; no stdout parsing.

### Routes

- `POST /api/memory/runtime/rebuild` accepts exactly `{"confirm": true}`; reuses existing CSRF / signed-user-proof chain; no config or force fields.
- Settings-driven confirmed path schedules the same `MemoryRuntime.rebuild()` after durable candidate+marker write.
- Manual “Rebuild index” without a settings change is **out of scope for this issue** (can follow once the marker/retry path is solid). Retry while the marker is already pending is in scope.

### Minimal Settings UI (ships with this issue)

Necessary UI only — not a horizontal recovery-UI bucket:

- Identity-changing Save opens a non-destructive confirmation (cost/duration disclosure); draft retained and same patch replayed with `confirm_rebuild: true`.
- Embedding identity fields remain editable when data exists (remove the data-exists lock).
- Persistent `rebuild_required` warning while marker pending.
- Footer: `Retry rebuild` while marker pending (beside existing Clear all). No permanent extra Rebuild control required beyond retry under pending marker.
- `Restart engine` disabled while rebuild required or rebuild task running.
- All new copy through `ui/src/i18n/en.json` and `zh.json`.

Processing Record state rows, dual-warning polish beyond what health already shows, Dependencies-page work, and full recovery-ladder user docs are **not** in this issue.

### Error codes (public, i18n-cataloged)

- `memory_embedding_rebuild_required`
- `memory_rebuild_failed`
- `memory_rebuild_root_busy`
- `memory_operation_in_progress` (or document reuse of an existing conflict code if one already gates all Memory mutations)

## Testing Decisions

Hermetic only: test-owned homes, fake artifacts/interpreters, no network, no local `vibe` service, no production paths.

1. **UI memory routes**: confirmed PATCH, 409 without confirm, exact rebuild body, CSRF/signed proof, join-on-duplicate, final-result responses (not 202), no secrets in payloads, cross-process/stale-save interleaving against a fake dual-writer or sequenced Controller write-back.
2. **MemoryRuntime layer**: quiesce failure, validation-before-destructive-stop, strict-empty vs indeterminate, result→settlement mapping, marker clear then candidate-snapshot refresh before activation, boot/orphan matrix, disabled-candidate empty settlement.
3. **Focused UI tests** where the existing frontend harness already supports them: draft retention + confirm replay, retry label under pending marker, restart disabled while pending.

Assert external behavior only (responses, persisted marker/candidate, child argv/env, process-group death, projection fields) — never internal call order or stdout.

Acceptance criteria in this issue are **self-contained**. Untracked companion plans under `docs/plans/` or `docs/memory-*.md` are background only and are **not** acceptance contracts. Do not require “plan test group N”.

## Out of Scope

- Factory reset (#1315).
- Generic Controller operation framework, 202 Accepted, client polling, operation IDs/history/progress, reserved `factory_reset` operation kind.
- Manual rebuild without a pending marker.
- Processing Record operation row, Dependencies-page UI, full recovery-ladder user documentation.
- Repair index / `cascade sync` / `fix --apply` (#1318).
- Release workflow / multi-platform artifact publish (already present in `.github/workflows/publish.yml`; not implementation scope here).
- Force restart, automatic rebuild from health, `cascade backfill`, stdout parsing, EverOS source changes, snapshot format changes.

## Dependencies

- Requires merged #1307 and #1310 on `dev`.
- Base PR on latest `origin/dev`.

### PR DESCRIPTION
**Pull request by rkrkrkk**
## Capability

Adds a confirmed Memory embedding rebuild recovery path. Embedding identity changes persist a durable rebuild intent before destructive work, run through one retained runtime task with cross-process admission, fence claims across failure/restart, and expose a truthful confirmation/Retry flow in Settings.

Closes #1314

## Scenarios

- `MEMORY-REBUILD-001`: confirmed first or changed embedding identity reaches a closed rebuild result
- `MEMORY-REBUILD-101`: unconfirmed identity writes are rejected without persistence
- `MEMORY-REBUILD-201`: failed rebuild retains the candidate and intent for Retry

## Evidence

- **Unit:** config transaction and legacy-marker migration, runtime fencing/preflight/settlement, process-group shutdown, retained-task concurrency, operation lease, and UI state coverage
- **Contract:** signed internal client/server request shape, exact browser confirmation payload, child argv/result mapping, and config compare-and-set behavior
- **Scenario:** catalog and closed-loop route cases for `MEMORY-REBUILD-001`, `MEMORY-REBUILD-101`, and `MEMORY-REBUILD-201`

## Validation

- Full Python suite: `1307 passed, 1 skipped`
- Full UI suite: `139 passed`
- Latest shutdown-focused set: backend `434 passed`; UI `26 passed`
- UI build: passed
- UI lint: passed
- Changed Python Ruff: passed
- Diff check: passed

## Residual manual checks

- Real pinned EverOS rebuild was not run.
- Incus regression was not run.
- Live process-group termination was not exercised outside hermetic tests.
- The managed-runtime artifact contract test is skipped because the managed runtime is not installed in this worktree.

## Scope

No `202`/polling API, generic operation framework, factory reset, `cascade sync`, or `fix --apply` behavior is included.

### CODE REVIEW

**Comment by cyhhao** [comment]
@codex review

File: core/memory/runtime.py [hunk 1]
```diff
@@ -1633,17 +1732,316 @@ async def _restart_locked(self) -> dict[str, Any]:
...
+            except Exception:
+                self._runtime_error = "memory_rebuild_failed"
+                return {
+                    "ok": False,
+                    "error": "memory_rebuild_failed",
+                    "result": "failed",
+                }
+
+            if pre_stop_has_data:
+                if not _rebuild_settings_usable(rebuild_settings):
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply core/memory/runtime.py:1846]: Fixed in 679c93cd. MemoryRuntime now runs the existing candidate processing probe under the claim fence before stopping the worker or sidecar, including for an empty root; a failed probe returns memory_rebuild_failed while retaining the durable marker and old sidecar. Added test_runtime_rebuild_probes_candidate_before_destructive_stop and kept the closed-result/API-key race coverage green.

File: ui/src/components/settings/memory/MemorySettingsPanel.tsx [hunk 2]
```diff
@@ -265,28 +331,38 @@ export const MemorySettingsPanel: React.FC<{
         draft={embeddingDraft}
         original={settings.processing.embedding}
         onChange={setEmbeddingDraft}
-        disabled={saving}
-        locked={embeddingLocked}
-        // Distinguish the two lock reasons: data-exists (permanent until Clear all) vs maintenance
-        // not yet resolved (transient — re-enables once the local fact confirms no data exists).
-        lockedHint={embeddingLockedHint}
+        disabled={busy}
+        identityDisabled={rebuildRequired}
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply ui/src/components/settings/memory/MemorySettingsPanel.tsx]: Fixed in 679c93cd. The pending marker no longer disables the embedding base URL/model fields; saving a corrected identity still uses the existing confirmation replay with confirm_rebuild: true, while API-key-only correction remains unchanged. Added component coverage for pending identity correction and a route contract test proving durable candidate replacement plus closed rebuild failure semantics.

**Comment by cyhhao** [comment]
@codex review

File: ui/src/components/settings/memory/MemorySettingsPanel.tsx [hunk 3]
```diff
@@ -255,8 +321,8 @@ export const MemorySettingsPanel: React.FC<{
         draft={llmDraft}
         original={settings.processing.llm}
         onChange={setLlmDraft}
-        disabled={saving}
-        locked={false}
+        disabled={busy}
+        identityDisabled={rebuildRequired}
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply ui/src/components/settings/memory/MemorySettingsPanel.tsx]: Fixed in 453489e6. The pending rebuild marker no longer disables LLM base URL/model fields; the shared endpoint identity lock prop was removed, while request-level rebuild busy still disables all settings. Added component coverage that corrects both LLM identity fields under rebuild_required and verifies the ordinary settings patch is submitted without embedding rebuild confirmation; the existing route contract for pending-marker LLM reconciliation also passes.

File: core/memory/runtime.py [hunk 4]
```diff
@@ -1269,6 +1294,8 @@ async def restore_backup(
     async def clear(self, *, operator_ref: str) -> dict[str, Any]:
         if not self.available:
             raise self._unavailable()
+        if self._rebuild_running():
+            return {"status": "failed", "error": "memory_operation_in_progress"}
```
[the remark this answers is not in the document]
Comment (rkrkrkk) [reply core/memory/runtime.py]: Fixed in 453489e6. MemoryRuntime now acquires the existing cross-process MemoryOperationLease around clear, resume_clear, and abort_clear, preserving the local rebuild fast-fail and returning memory_operation_in_progress when another controller owns the lease. Extended the two-controller rebuild lease regression to prove all three destructive clear paths are rejected while the first controller rebuild is active; runtime, maintenance, lock, route, config, and process suites pass.

**Comment by cyhhao** [comment]
@codex review
